### PR TITLE
Added prefetcher to decoder

### PR DESCRIPTION
**Pull request by levBagryansky**

### CODE REVIEW

**Review by mmamayka — COMMENTED** [review]
One little fix and it will be awesome

File: include/besm-666/decoder/prefetcher.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Virtual address that corresponds to {@link Prefetcher::saved_}.
+     */
+    RV64Ptr start_;
+
+    /**
+     * Number of continuous bytes starting from {@link Prefetcher::start_}
+     */
+    size_t len_;
```
[the remark this answers is not in the document]
Comment (levBagryansky) [reply include/besm-666/decoder/prefetcher.hpp]: @mmamayka done

File: include/besm-666/memory/mmu.hpp [hunk 2]
```diff
@@ -24,8 +24,10 @@ class MMU : public INonCopyable {
     void storeWord(RV64Ptr address, RV64Word value);
     void storeDWord(RV64Ptr address, RV64DWord value);
 
-    std::pair<void *, size_t> getHostAddress(RV64Ptr address);
-    std::pair<void const *, size_t> getHostAddress(RV64Ptr address) const;
+    std::pair<void *, size_t> touchHostAddress(RV64Ptr vaddress);
```
Comment (ivan-egorov42) [inline include/besm-666/memory/mmu.hpp:27]: Mb need `RV64Size` too? Also, where is an implementation?

File: src/decoder/prefetcher.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace besm::dec {
+
+Prefetcher::Prefetcher(mem::MMU::SPtr mmu)
+    : mmu_(std::move(mmu)), saved_(nullptr), start_(-1), len_(0) {}
+
+RV64UWord Prefetcher::loadWord(RV64Ptr vaddress) {
+    if (vaddress > start_ && vaddress < start_ + len_) {
+        // this address was already load
+        return *(RV64UWord *)((uint8_t *)saved_ + vaddress - start_);
```
Comment (ivan-egorov42) [inline src/decoder/prefetcher.cpp:11]: Mb using static_cast better?

File: include/besm-666/memory/mmu.hpp (hunk 2 above)
Comment (levBagryansky) [reply include/besm-666/memory/mmu.hpp:27]: @ivan-egorov42 I just renamed it here

File: src/decoder/prefetcher.cpp (hunk 3 above)
Comment (levBagryansky) [reply src/decoder/prefetcher.cpp]: @ivan-egorov42 done

File: include/besm-666/memory/mmu.hpp (hunk 2 above)
Comment (levBagryansky) [reply include/besm-666/memory/mmu.hpp]: @ivan-egorov42 made RV64Size here

File: src/sim/hart.cpp [hunk 4]
```diff
@@ -14,8 +15,9 @@ Hart::SPtr Hart::Create(std::shared_ptr<mem::PhysMem> const &pMem,
 }
 
 Hart::Hart(std::shared_ptr<mem::PhysMem> const &pMem,
```
Comment (c71n93) [inline src/sim/hart.cpp:17]: Could you return it back, please? New version does not optimize anything. It just moves copying of `hookManager` from initializer list to call of `Hart::Hart` constructor.

**Review by c71n93 — CHANGES_REQUESTED** [review]

File: src/sim/hart.cpp (hunk 4 above)
Comment (levBagryansky) [reply src/sim/hart.cpp:17]: @c71n93 no, this is not worse but maybe better since clang-tidy recommend it

Comment (c71n93) [reply src/sim/hart.cpp:17]: @levBagryansky prove that it's better, and then you can leave

**Review by c71n93 — APPROVED** [review]

File: src/sim/hart.cpp (hunk 4 above)
Comment (levBagryansky) [reply src/sim/hart.cpp:17]: @c71n93 this is not worse but maybe better since clang-tidy recommend it

**Comment by levBagryansky** [comment]
@ivan-egorov42  please check

**Review by ivan-egorov42 — APPROVED** [review]
LGTM

**Pull request merged**